### PR TITLE
Sabre layout and routing transpiler passes

### PR DESCRIPTION
**Pull request by ajavadia**
Adding a layout and routing method from https://arxiv.org/abs/1809.02573 (by @Skywalker2012).

The initial layout selection is especially good in this method and finds high-quality initial placements very fast. I allow this layout method to be combined with any of the other available routing methods as well through selection of `layout_method` and `routing_method`. But Sabre's own routing method is also quite good - it anticipates upcoming gates via look ahead, and gives you the ability to trade off parallelism vs. number of gates.

I have not made this part of the default pipeline yet, but I think I'll do that once I do more randomized testing.

### CODE REVIEW

**Comment by Skywalker2012** [comment]
It looks good to me.

File: qiskit/circuit/bit.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ def index(self, value):
 
     def __repr__(self):
         """Return the official string representing the bit."""
-        return "%s(%s, %s)" % (self.__class__.__name__, self._register, self._index)
+        return "%s[%s]" % (self._register.name, self._index)
```
Comment (1ucian0) [inline qiskit/circuit/bit.py:73]: Why is this necessary?
```suggestion
        return "%s(%s, %s)" % (self.__class__.__name__, self._register, self._index)
```

File: qiskit/dagcircuit/dagcircuit.py [hunk 2]
```diff
@@ -1097,6 +1097,16 @@ def predecessors(self, node):
         """Returns iterator of the predecessors of a node as DAGNodes."""
         raise NotImplementedError()
 
+    def quantum_successors(self, node):
```
Comment (1ucian0) [inline qiskit/dagcircuit/dagcircuit.py:1100]: Since this extends the current `DAGCircuit` API, I suggest to add a test in `test/python//dagcircuit/test_dagcircuit.py`

File: qiskit/dagcircuit/dagcircuit.py [hunk 3]
```diff
@@ -1182,6 +1182,19 @@ def remove_nondescendants_of(self, node):
             if n.type == "op":
                 self.remove_op_node(n)
 
+    def front_layer(self):
```
Comment (1ucian0) [inline qiskit/dagcircuit/dagcircuit.py:1185]: Since this extends `DAGCircuit` API, maybe a test in `test/python//dagcircuit/test_dagcircuit.py`

File: qiskit/transpiler/coupling.py [hunk 4]
```diff
@@ -201,6 +206,17 @@ def is_symmetric(self):
             self._is_symmetric = self._check_symmetry()
         return self._is_symmetric
 
+    def make_symmetric(self):
```
Comment (1ucian0) [inline qiskit/transpiler/coupling.py:209]: I added a test for this in e4e42ed65

File: qiskit/transpiler/coupling.py [hunk 5]
```diff
@@ -134,6 +134,11 @@ def is_connected(self):
         except nx.exception.NetworkXException:
             return False
 
+    def neighbors(self, physical_qubit):
```
Comment (1ucian0) [inline qiskit/transpiler/coupling.py:137]: I'm not sure if this behaves as expected:
```python
from qiskit.transpiler import CouplingMap
coupling_list = [[0, 1], [0, 2]]
coupling = CouplingMap(coupling_list)
physical_qubits = coupling.physical_qubits

print([qubit for qubit in coupling.neighbors(physical_qubits[1])])
```
```
[]
```
Shouldn't this be `[0]`?

Comment (1ucian0) [reply qiskit/transpiler/coupling.py:137]: I suggest a test like this in `test/python/transpiler/test_coupling.py`: 
```python 

        coupling_list = [[0, 1], [0, 2]]
        coupling = CouplingMap(coupling_list)

        physical_qubits = coupling.physical_qubits
        self.assertEqual(set(coupling.neighbors(physical_qubits[0])), set([1, 2]))
```

File: qiskit/transpiler/layout.py [hunk 6]
```diff
@@ -224,10 +223,6 @@ def combine_into_edge_map(self, another_layout):
             LayoutError: another_layout can be bigger than self, but not smaller.
                 Otherwise, raises.
         """
-        warnings.warn('combine_into_edge_map is deprecated as of 0.14.0 and '
-                      'will be removed in a future release. Instead '
-                      'reorder_bits() should be used', DeprecationWarning,
-                      stacklevel=2)
```
Comment (1ucian0) [inline qiskit/transpiler/layout.py:230]: This means that `combine_into_edge_map` is not going to be removed in the future?

File: qiskit/transpiler/passes/layout/sabre_layout.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        Raises:
+            TranspilerError: if dag wider than self.coupling_map
+        """
+        from qiskit.converters import dag_to_circuit
+        from qiskit.transpiler.passes.layout.set_layout import SetLayout
+        from qiskit.transpiler.passes.layout.full_ancilla_allocation import FullAncillaAllocation
+        from qiskit.transpiler.passes.layout.enlarge_with_ancilla import EnlargeWithAncilla
+        from qiskit.transpiler.passes.layout.apply_layout import ApplyLayout
+        from qiskit.transpiler.passes.routing import SabreSwap
+        from qiskit.transpiler.passmanager import PassManager
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/sabre_layout.py:78]: why importing at runtime?

File: qiskit/transpiler/passes/layout/sabre_layout.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                                     len(dag.qubits), replace=False)
+        physical_qubits = rng.permutation(physical_qubits)
+        initial_layout = Layout({q: dag.qubits[i]
+                                 for i, q in enumerate(physical_qubits)})
+
+        if self.routing_pass is None:
+            self.routing_pass = SabreSwap(self.coupling_map, 'decay')
+
+        # Do forward-backward iterations.
+        circ = dag_to_circuit(dag)
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/sabre_layout.py:128]: This conversion might be avoid by calling the passes directly instead of using a passmanager.

File: qiskit/transpiler/passes/layout/sabre_layout.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        from qiskit.transpiler.passes.layout.apply_layout import ApplyLayout
+        from qiskit.transpiler.passes.routing import SabreSwap
+        from qiskit.transpiler.passmanager import PassManager
+
+        def _layout_and_route_passmanager():
+            """Return a passmanager for a full layout and routing.
+
+            We use a factory to remove potential statefulness of passes.
+            """
+            layout_and_route = [SetLayout(initial_layout),
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/sabre_layout.py:85]: considering passing this `initial_layout` as a parameter to the function. Otherwise, the scope gets confusing.

File: qiskit/circuit/bit.py (hunk 1 above)
Comment (ajavadia) [reply qiskit/circuit/bit.py]: it was for my debugging.

File: qiskit/transpiler/layout.py (hunk 6 above)
Comment (ajavadia) [reply qiskit/transpiler/layout.py:230]: yeah i restored it

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 7 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/layout/sabre_layout.py]: fixed

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 8 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/layout/sabre_layout.py:127]: I need the circuit.reverse_ops() method which is not implemented for DAGs.

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 9 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/layout/sabre_layout.py:84]: done thanks

File: qiskit/transpiler/coupling.py (hunk 5 above)
Comment (ajavadia) [reply qiskit/transpiler/coupling.py:137]: This is because the graph is directional. If you make it bidirectional then it will be a neighbor. I added a note and and a test.

Comment (1ucian0) [reply qiskit/transpiler/coupling.py:137]: thanks! Adjusting and adding the test in 4b3a65b1a

File: qiskit/transpiler/layout.py (hunk 6 above)
Comment (1ucian0) [reply qiskit/transpiler/layout.py:230]: That might need a release note for reversing the deprecation note from the past, I guess. Not sure what's the expected thing here.

File: qiskit/dagcircuit/dagcircuit.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/dagcircuit/dagcircuit.py:1185]: Test added in 5527c0512

File: qiskit/transpiler/passes/layout/sabre_layout.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    def run(self, dag):
+        """Run the SabreLayout pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): DAG to find layout for.
+
+        Raises:
+            TranspilerError: if dag wider than self.coupling_map
+        """
+        def _layout_and_route_passmanager(initial_layout):
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/sabre_layout.py:79]: It seems to me that this nested function can be a method.

File: qiskit/transpiler/passes/layout/sabre_layout.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+            """
+            layout_and_route = [SetLayout(initial_layout),
+                                FullAncillaAllocation(self.coupling_map),
+                                EnlargeWithAncilla(),
+                                ApplyLayout(),
+                                self.routing_pass]
+            pm = PassManager(layout_and_route)
+            return pm
+
+        def _compose_layouts(initial_layout, pass_final_layout, qregs):
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/sabre_layout.py:92]: [Similarly](https://github.com/Qiskit/qiskit-terra/pull/4537/files#r443141810), this one.

File: qiskit/transpiler/passes/layout/sabre_layout.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                            for v, _ in initial_layout.get_virtual_bits().items()}
+            return Layout(final_layout)
+
+        if len(dag.qubits) > self.coupling_map.size():
+            raise TranspilerError('More virtual qubits exist than physical.')
+
+        # Choose a random initial_layout.
+        if self.seed is None:
+            self.seed = np.random.randint(0, np.iinfo(np.int32).max)
+        rng = np.random.default_rng(self.seed)
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/sabre_layout.py:115]:
```suggestion
        rng = np.random.RandomState(self.seed)
```

File: qiskit/transpiler/passes/routing/sabre_swap.py [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+            TranspilerError: if the coupling map or the layout are not
+            compatible with the DAG
+        """
+        if len(dag.qregs) != 1 or dag.qregs.get('q', None) is None:
+            raise TranspilerError('Sabre swap runs on physical circuits only.')
+
+        if len(dag.qubits) > self.coupling_map.size():
+            raise TranspilerError('More virtual qubits exist than physical.')
+
+        rng = np.random.default_rng(self.seed)
```
Comment (1ucian0) [inline qiskit/transpiler/passes/routing/sabre_swap.py:147]:
```suggestion
        rng = np.random.RandomState(self.seed)
```

File: test/python/transpiler/test_sabre_swap.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline test/python/transpiler/test_sabre_swap.py:5]:
```suggestion
# (C) Copyright IBM 2017, 2020.
```

File: qiskit/transpiler/passes/routing/sabre_swap.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+
+        # Assume bidirectional couplings, fixing gate direction is easy later.
+        self.coupling_map.make_symmetric()
+
+        canonical_register = dag.qregs['q']
+        current_layout = Layout.generate_trivial_layout(canonical_register)
+
+        # A decay factor for each qubit used to heuristically penalize recently
+        # used qubits (to encourage parallelism).
+        self.qubits_decay = dict(zip(dag.qubits, len(dag.qubits) * [1]))
```
Comment (1ucian0) [inline qiskit/transpiler/passes/routing/sabre_swap.py:160]:
```suggestion
        self.qubits_decay = {qubit: 1 for qubit in dag.qubits}
```

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 12 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/layout/sabre_layout.py:115]: no this needs to be like this i keep saying this!
https://github.com/Qiskit/qiskit-terra/pull/3586#discussion_r421211627

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 10 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/layout/sabre_layout.py:79]: i don't think so, it's a locally used function, not sure why a method is better.

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 11 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/layout/sabre_layout.py:92]: same

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 12 above)
Comment (1ucian0) [reply qiskit/transpiler/passes/layout/sabre_layout.py:115]: I keep missing them up. Sorry.
```suggestion
        rng = np.random.default_rng(self.seed)
```

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 10 above)
Comment (1ucian0) [reply qiskit/transpiler/passes/layout/sabre_layout.py:79]: Cleans `run` and, because the scope is more limited, it is easy to read, avoiding snicky issues.

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 11 above)
Comment (1ucian0) [reply qiskit/transpiler/passes/layout/sabre_layout.py:92]: Notice that `_obtain_swaps` does not follow this pattern.

File: qiskit/transpiler/passes/routing/sabre_swap.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+                score = self._score_heuristic(self.heuristic,
+                                              front_layer,
+                                              extended_set,
+                                              trial_layout,
+                                              swap_qubits)
+                swap_scores[swap_qubits] = score
+            min_score = min(swap_scores.values())
+            best_swaps = [k for k, v in swap_scores.items() if v == min_score]
+            best_swaps.sort(key=lambda x: (x[0].index, x[1].index))
+            best_swap = rng.choice(best_swaps)
```
Comment (1ucian0) [inline qiskit/transpiler/passes/routing/sabre_swap.py:223]: My understanding of `rng.choice` is that `best_swaps` will be shuffled.

```suggestion
            best_swap = rng.choice(best_swaps)
```

File: qiskit/transpiler/passes/routing/sabre_swap.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+
+        elif heuristic == 'decay':
+            return max(self.qubits_decay[swap_qubits[0]], self.qubits_decay[swap_qubits[1]]) * \
+                   self._score_heuristic('lookahead', front_layer, extended_set, layout)
+
+        else:
+            raise TranspilerError('Heuristic %s not recognized.' % heuristic)
+
+
+def _copy_circuit_metadata(source_dag):
```
Comment (1ucian0) [inline qiskit/transpiler/passes/routing/sabre_swap.py:340]: from the description, I suggest the name `_copy_dag_metadata`

File: qiskit/transpiler/passes/routing/sabre_swap.py (hunk 16 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/routing/sabre_swap.py:223]: that's fine, but i need to sort it so that a given seed behaves deterministically. otherwise it gives a different result each time (i tried it).

File: qiskit/transpiler/passes/routing/sabre_swap.py (hunk 17 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/routing/sabre_swap.py:340]: i copied this from the lookahead_swap pass and dag and circuit are basically the same thing

File: qiskit/transpiler/passes/layout/sabre_layout.py (hunk 10 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/layout/sabre_layout.py:79]: ok done

File: qiskit/transpiler/layout.py (hunk 6 above)
Comment (mtreinish) [reply qiskit/transpiler/layout.py:230]: Yeah, I would add a release note here probably in the other section (it doesn't really fit into the upgrade or deprecation sections). Something like:

```rst
The layout method :meth:`qiskit.transpiler.Layout.combine_into_edge` is no longer deprecated. It was deprecated
in the 0.14.0 release because the functionality overlapped with the :meth:`qiskit.transpiler.Layout.reorder_bits`
method but there were use cases for the method over what ``reorder_bits()`` provided so it is no longer deprecated,
 and will no longer be removed from a future release.
```

Comment (ajavadia) [reply qiskit/transpiler/layout.py:230]: added
